File: mario-less/credit.c
#include <stdio.h>
#include <cs50.h>

int get_digit_sum(int number);
string validate_credit_card(long long card_number);

int main(void)
{
    // Get credit card number from the user
    long long card_number;
    do
    {
        card_number = get_long("Number: ");
    } while (card_number <= 0);

    // Validate and identify the card type
    string result = validate_credit_card(card_number);

    // Print the result
    printf("%s\n", result);

    return 0;
}

int get_digit_sum(int number)
{
    // Calculate the sum of the digits of a number
    int sum = 0;
    while (number > 0)
    {
        sum += number % 10;
        number /= 10;
    }
    return sum;
}

string validate_credit_card(long long card_number)
{
    int digit_sum = 0;
    int multiplier = 1;

    // Iterate through the digits of the credit card number
    while (card_number > 0)
    {
        int digit = card_number % 10;

        // Multiply every other digit by 2
        if (multiplier % 2 == 0)
        {
            digit *= 2;
            digit_sum += get_digit_sum(digit);
        }
        else
        {
            digit_sum += digit;
        }

        // Move to the next digit
        card_number /= 10;
        multiplier++;
    }

    // Check if the card number is valid
    if (digit_sum % 10 == 0)
    {
        // Identify the card type based on starting digits
        if ((card_number == 34 || card_number == 37) && multiplier == 15)
        {
            return "AMEX";
        }
        else if ((card_number >= 51 && card_number <= 55) && multiplier == 16)
        {
            return "MASTERCARD";
        }
        else if ((card_number / 10 == 4 || card_number / 1000 == 4) && (multiplier == 13 || multiplier == 16))
        {
            return "VISA";
        }
    }

    // If not valid or card type not identified, return INVALID
    return "INVALID";
}

File: sentimental-credit/credit.py
from cs50 import get_int

# TODO

# prompt user for input
while True:
    card = get_int("Card: ")

    if card > 0:
        break




def luhn_checksum(card):
    def digits_of(n):
        return [int(d) for d in str(n)]
    digits = digits_of(card)
    even_digits = digits[-2::-2]
    odd_digits = digits[-1::-2]
    checksum = digits[-1]
    for d in even_digits:
        checksum += sum(digits_of(d*2))
    return checksum % 10


length = len(str(card))


if luhn_checksum(card) == 0:
    if length == 13 or length == 16:
        if str(card)[0] == '4':
            print("Visa")
        elif str(card)[0:2] == '51' or str(card)[0:2] == '52' or str(card)[0:2] == '53' or str(card)[0:2] == '54' or str(card)[0:2] == '55':
            print("Master")
    elif length == 15:
        if str(card)[0:2] == '34' or str(card)[0:2] == '37':
            print("Amex")
    else:
        print("Invalid card")
else:
    print("Invalid")
